### PR TITLE
Add NB for BERT models

### LINKED ISSUE #81: Add a tutorial on how to use Huggingface models with giotto-deep
It would be nice to have a tutorial on how to use pre-trained Huggingface transformer models with giotto-deep. 
This would show how to fine-tune within giotto-deep and how to use some interpretability methods on them.
This would replace the existing notebooks on QA and translation which use a T5-style model with random initialization.

### PR DESCRIPTION
**Pull request by hkirvesl**
**Reference issues/PRs**

Fixes #81 By adding a notebook on BERT models with topological interpretability methods built on top of one. The old tutorials on Q&A and Translation remain for now.

**Types of changes**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Description**

1) Add notebook for training HuggingFace Bert model with giotto-deep, and how to use topological pruning methods for such models.
2) Add support for training HuggingFace models with an unwrapper in Trainer.py
3) Modify models/utils.py to support the the HuggingFace models (now with an unwrapper)

**Screenshots (if appropriate)**

**Any other comments?**

**Checklist**

- [x] I have read the [guidelines for contributing](https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines).
- [x] My code follows the code style of this project. I used `flake8` to check my Python changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed. I used `pytest` to check this on Python tests.

### CODE REVIEW

File: examples/basic_tutorial_BERT.ipynb [hunk 1]
```diff
@@ -0,0 +1,1172 @@
...
+    "3. Fit more complicated models;\n",
+    "4. Optimize the parameters.\n"
+   ]
+  }
+ ],
+ "metadata": {
+  "kernelspec": {
+   "display_name": "Python [conda env:torch] *",
+   "language": "python",
+   "name": "conda-env-torch-py"
```
Comment (matteocao) [inline examples/basic_tutorial_BERT.ipynb:1155]: the CI is complaining about this name: 

```
jupyter_client.kernelspec.NoSuchKernel: No such kernel named conda-env-torch-py
```

Can you please change this to:
```
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
```

**Review by matteocao — CHANGES_REQUESTED** [review]
Once the CI pass, I will study your notebook in details, as I am really curious about it!

File: examples/basic_tutorial_BERT.ipynb (hunk 1 above)
Comment (hkirvesl) [reply examples/basic_tutorial_BERT.ipynb]: Thank you! Should be fixed now

Comment (matteocao) [reply examples/basic_tutorial_BERT.ipynb]: @hkirvesl can you please use `os.path.join` to build file path please? windows and posix OS have a different convention (back and forward slashes)

Screen Shot 2022-09-29 at 19 56 54

Comment (hkirvesl) [reply examples/basic_tutorial_BERT.ipynb]: Thank you for the suggestion. Done.

File: gdeep/trainer/trainer.py [hunk 2]
```diff
@@ -266,9 +266,13 @@ def _send_to_device(
                 new_x.append(xi.to(DEVICE))
             x = new_x
             prediction = self.model(*x)
+            if(hasattr(prediction, "logits")): # unwrapper for HuggingFace BERT model
+                prediction = prediction.logits # unwrapper for HuggingFace BERT model
```
Comment (matteocao) [inline gdeep/trainer/trainer.py:270]: all good here!

**Review by matteocao — CHANGES_REQUESTED** [review]
Dear @hkirvesl ,

I think we are almost there: I only have a few minor comments.
I will then review the notebbook in detail.

**One important note:** did you pull master to your fork before committing? I ask because there is one item in the CI that took > 5h to run and has currently not yet finished (though it should really take 25 mins at most) and I wonder if it is related to not having integrated the fixes. If you have done so, than ignore this last comment

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -6,11 +6,11 @@ repos:
...
-    hooks:
-    -   id: black
--   repo: https://github.com/pre-commit/mirrors-mypy
-    rev: 'v0.971'  # Use the sha / tag you want to point at
-    hooks:
-    -   id: mypy
+#-   repo: https://github.com/psf/black
+#    rev: 22.8.0
+#    hooks:
+#    -   id: black
```
[the remark this answers is not in the document]
Comment (raphaelreinauer) [reply .pre-commit-config.yaml:12]: Could you please wait until my pull request is accepted and then pull again from master? I fixed all mypy errors, and the pre-commit does work now.

File: examples/basic_tutorial_BERT.ipynb [hunk 4]
```diff
@@ -0,0 +1,1198 @@
+{
```
Comment (matteocao) [inline examples/basic_tutorial_BERT.ipynb:1]: <p>here you can simply use</p><pre class="ql-syntax" spellcheck="false">from gdeep.utility import DEVICE
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (matteocao) [inline examples/basic_tutorial_BERT.ipynb:1]: <p>ca you please preceed this cell with a markdown cell describing what you are doing here?</p><p> </p><p>if this cell and the above one are related to handling masks, then maybe it is worth putting a bit of explanation 2 cells above, details the steps that one needs to take</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (matteocao) [inline examples/basic_tutorial_BERT.ipynb:1]: <p>can you please add a few inline comments to describe what you are doing here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (matteocao) [inline examples/basic_tutorial_BERT.ipynb:1]: <p>the comment here , I think, needs to be corrected: <code>raw</code> -&gt; <code>mean aggregated</code>   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (matteocao) [inline examples/basic_tutorial_BERT.ipynb:1]: <p>some more comments here please</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (matteocao) [inline examples/basic_tutorial_BERT.ipynb:1]: <p>if these two cells are "almost" duplicated, can you please create a function in one cell and call it with different parameters?</p><p> </p><p>Probably same comments may work a few cells above</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (matteocao) [inline examples/basic_tutorial_BERT.ipynb:1]: <p>typo: At such --&gt; As such</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (matteocao) [inline examples/basic_tutorial_BERT.ipynb:1]: <p>maybe one sentence describe the two cells below.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='43'/>

Comment (matteocao) [inline examples/basic_tutorial_BERT.ipynb:1]: <p>ca you clarify why you compute the gradients (and of what) in a sentence in a .md cell above this one?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='54'/>

**Comment by hkirvesl** [comment]
Thanks for the good comments! I pulled the main and it did fix a lot, (and also managed to close this PR.) Will reopen once I've cleaned the notebook more a bit more as per the suggestions

**Comment by raphaelreinauer** [comment]
Could you please install pre-commit (see README.md)

**Comment by hkirvesl** [comment]
Reopening the PR

Changelog:

**The notebook:**

1) Imports

1.1) Removed unneeded imports:
from gdeep.visualisation import persistence_diagrams_of_activations
from gdeep.visualisation import Visualiser

1.2) Added import

from gdeep.utility import DEVICE

2) Cleaned some variable names

3) Added more explanation as per the suggested comments

4) Removed unnecessary code chunks

5) renamed variables

6) Added functions to reduce redundant code

7) Add legend to a persistence diagram plot.

**Testing:**

1) Run all the prehooks, both black and mypy

**models/utils.py**

1) Replace the modification to retain the existing functionality

File: .pre-commit-config.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (hkirvesl) [reply .pre-commit-config.yaml]: Thanks for the comments, Matteo and Raphael. I did recall some discussion about problems with the prehooks and commenting them out being the recommended temporary work around. I just forgot the details and ended up commenting out too much. All these should be fixed now. Thank you Raphael for the fix, this makes our lives much easier.

File: gdeep/models/utils.py [hunk 5]
```diff
@@ -70,6 +70,6 @@ def __call__(
                 the output tensor of the module
         """
         if isinstance(module_out, tuple):
-            self.outputs.append(module_out[0].detach())
+            self.outputs.append(module_out[0])#.detach())
```
[the remark this answers is not in the document]
Comment (hkirvesl) [reply gdeep/models/utils.py]: Thanks for the comment, this is a very good point indeed. I do understand the consequences and I agree what was proposed is not a good solution. To be honest, I forgot about this myself. It was just a quick and dirty fix to make the BERT models work and since it worked and seemingly didn't break any tests I just proceeded without worrying about it.

A good solution would keep the old functionality where applicable, in the same spirit as the change in trainer.py. I believe this has now been fixed.

File: examples/basic_tutorial_BERT.ipynb (hunk 4 above)
Comment (hkirvesl) [reply examples/basic_tutorial_BERT.ipynb:1]: Thanks! It is a more giotto-way to do this. In general, any other comments on how to make the notebook more giotto-like are  highly appreciated.

Comment (hkirvesl) [reply examples/basic_tutorial_BERT.ipynb:1]: <p>Good idea.I added a brief note (and deleted an unnecessary chunk)</p>

Comment (hkirvesl) [reply examples/basic_tutorial_BERT.ipynb:1]: <p>You are absolutely right. Fixed.</p>

Comment (hkirvesl) [reply examples/basic_tutorial_BERT.ipynb:1]: <p>Good catch!</p>

Comment (hkirvesl) [reply examples/basic_tutorial_BERT.ipynb:1]: <p>Added some discussion on this. Also: If you know of a more giotto-like way of getting your hands on these gradients I think it would improve this notebook.</p>

**Review by matteocao — APPROVED** [review]
this loks good and it is very interesting btw! thanks @hkirvesl !!
